node/pkg/watchers/sui/watcher.go

```diff
@@ -503,25 +443,146 @@ func (e *Watcher) Run(ctx context.Context) error {
 
 	select {
 	case <-ctx.Done():
-		_ = ws.Close(websocket.StatusNormalClosure, "")
 		return ctx.Err()
 	case err := <-errC:
-		_ = ws.Close(websocket.StatusInternalError, err.Error())
 		return err
 	}
 }
 
-// fixSuiWsURL ensures the websocket scheme is properly specified
-func (e *Watcher) fixSuiWsURL(logger *zap.Logger) error {
-	u, _ := url.Parse(e.suiWS)
+func (w *Watcher) getEvents() (SuiEventResponseData, error) {
+	// Only get events newer than the last processed height
+	var retVal SuiEventResponseData
+	var nextCursor struct {
+		TxDigest string
+		EventSeq string
+	}
+	firstTime := true
+	done := false
+	for !done {
+		var reader io.Reader
+		if firstTime {
+			reader = strings.NewReader(
+				fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, null, %d, %t]}`,
+					w.suiMoveEventType, w.maximumBatchSize, w.descendingOrder))
+		} else {
+			reader = strings.NewReader(
+				fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, { "txDigest": "%s", "eventSeq": "%s" }, %d, %t]}`,
+					w.suiMoveEventType, nextCursor.TxDigest, nextCursor.EventSeq, w.maximumBatchSize, w.descendingOrder))
+		}
+		resp, err := http.Post(w.suiRPC, "application/json", reader) //nolint:noctx // TODO FIXME we should propagate context with Deadline here.
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to post: %w", err)
+		}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to read: %w", err)
+		}
+		resp.Body.Close()
+
+		var res SuiEventResponse
+		err = json.Unmarshal(body, &res)
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to unmarshal body: %s, error: %w", string(body), err)
+		}
+		for _, datum := range res.Result.Data {
+			// Check if the event is newer than the last processed height
+			height, hErr := w.getCheckpointForEvent(datum)
```

Comment:
Done is based on checkpoint.  Without actually getting the checkpoint for the event in question, how would I know if I am done?  Also, we are polling 3 times faster than blocks are being produced.  I think it is unlikely to get more than a few events of interest.  I'm happy to make this more efficient.  But I don't get what you are saying.

Reply:
On this point, I meant that instead of getting every event’s checkpoint in the first loop, you only need to get the _last_ event’s checkpoint since you are querying in descending order and that will be the oldest. That alone should be enough to calculate done, and gather all the applicable events. Then, after having gathered all the potentially applicable events, you can use the batch call to get all the checkpoints in one RPC call, then filter out the ones that are older than necessary.

Reply:
Changed